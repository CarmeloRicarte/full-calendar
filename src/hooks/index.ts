export * from "./dispatchSelector";
export * from "./useCalendarStore";
export * from "./useForm";
export * from "./useUiStore";
