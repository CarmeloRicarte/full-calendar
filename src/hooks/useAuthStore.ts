import { useSelector } from "react-redux";
import { calendarApi } from "../api";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
  RootState,
} from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector(
    (state: RootState) => state.auth
  );
  const dispatch = useAppDispatch();

  const startLogin = async ({
    email,
    password,
  }: {
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth", {
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error) {
      dispatch(onLogout("Credenciales incorrectas"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const startRegister = async ({
    name,
    email,
    password,
  }: {
    name: string;
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/new", {
        name,
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error: any) {
      if (error.response.data?.msg) {
        dispatch(onLogout(error.response.data?.msg));
      } else {
        const keyFirstElement = Object.keys(error.response.data?.errors)[0];
        keyFirstElement &&
          dispatch(onLogout(error.response.data?.errors[keyFirstElement]?.msg));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const setTokenLocalStorage = (token: string) => {
    localStorage.setItem("token", token);
    localStorage.setItem(
      "token-expires-date",
      new Date().getTime() + 2 * 60 * 60 * 1000 + ""
    );
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogout());
    try {
      const { data } = await calendarApi.get("/auth/renew");

      // logout user if token expired
      const expirationTokenDate = localStorage.getItem("token-expires-date");
      if (expirationTokenDate && +expirationTokenDate < new Date().getTime()) {
        dispatch(onLogout());
      } else {
        setTokenLocalStorage(data);
        dispatch(
          onLogin({
            name: data.name,
            uid: data.uid,
          })
        );
      }
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  return {
    //* Properties
    status,
    user,
    errorMessage,
    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
  };
};
