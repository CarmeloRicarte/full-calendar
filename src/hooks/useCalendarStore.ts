import { useAppDispatch } from "./dispatchSelector";
import { useSelector } from "react-redux";
import { onSetActiveEvent, RootState } from "../store";
import { CalendarEvent } from "../calendar/interfaces";

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useSelector(
    (state: RootState) => state.calendar
  );

  const setActiveEvent = (calendarEvent: CalendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  return {
    //* Properties
    events,
    activeEvent,
    //*Methods
    setActiveEvent,
  };
};
