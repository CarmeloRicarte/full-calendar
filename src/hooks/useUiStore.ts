import { useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal, RootState } from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useUiStore = () => {
  const dispatch = useAppDispatch();
  const { isDateModalOpen } = useSelector((state: RootState) => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => dispatch(onCloseDateModal());

  return {
    //* Properties
    isDateModalOpen,

    //*Methods
    openDateModal,
    closeDateModal,
  };
};
