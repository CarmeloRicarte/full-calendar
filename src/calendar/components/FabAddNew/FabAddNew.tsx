import { addHours } from "date-fns";
import { useCalendarStore, useUiStore } from "../../../hooks";
import "./FabAddNew.scss";
export const FabAddNew = () => {
  const { openDateModal } = useUiStore();
  const { setActiveEvent } = useCalendarStore();

  const handleClickNew = () => {
    setActiveEvent({
      title: "",
      notes: "",
      start: new Date(),
      end: addHours(new Date(), 2),
    });

    openDateModal();
  };
  return (
    <button onClick={handleClickNew} className="btn btn-primary fab">
      <i className="fas fa-plus"></i>
    </button>
  );
};
