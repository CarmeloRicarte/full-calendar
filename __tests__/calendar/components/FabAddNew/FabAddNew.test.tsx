import { describe, test, expect, vi, beforeEach } from "vitest";
import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { FabAddNew } from "../../../../src/calendar/components/FabAddNew/FabAddNew";
import * as useCalendarStore from "../../../../src/hooks/useCalendarStore";
import * as useUiStore from "../../../../src/hooks/useUiStore";

vi.mock("../../../../src/hooks/useCalendarStore");
vi.mock("../../../../src/hooks/useUiStore");

describe("Tests of <FabAddNew />", () => {
  const mockSetActiveEvent = vi.fn();
  const mockOpenDateModal = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should render button", () => {
    (useCalendarStore as any).useCalendarStore = vi.fn().mockReturnValue({
      setActiveEvent: mockSetActiveEvent,
    });

    (useUiStore as any).useUiStore = vi.fn().mockReturnValue({
      openDateModal: mockOpenDateModal,
    });
    const { container } = render(<FabAddNew />);
    expect(container).toMatchSnapshot();
  });

  test("should call setActiveEvent and openDateModal on click", () => {
    (useCalendarStore as any).useCalendarStore = vi.fn().mockReturnValue({
      setActiveEvent: mockSetActiveEvent,
    });

    (useUiStore as any).useUiStore = vi.fn().mockReturnValue({
      openDateModal: mockOpenDateModal,
    });
    render(<FabAddNew />);
    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(mockSetActiveEvent).toHaveBeenCalledWith({
      title: "",
      notes: "",
      start: expect.any(Date),
      end: expect.any(Date),
    });
    expect(mockOpenDateModal).toHaveBeenCalled();
  });
});
