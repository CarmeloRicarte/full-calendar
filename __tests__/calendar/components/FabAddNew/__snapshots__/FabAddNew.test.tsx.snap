// Vitest Snapshot v1

exports[`Tests of <FabAddNew /> > should render button 1`] = `
<div>
  <button
    class="btn btn-primary fab"
  >
    <i
      class="fas fa-plus"
    />
  </button>
</div>
`;
